### PR TITLE
feat: auto-populate metadata of log entries at write()

### PR DESCRIPTION
**Pull request by minherz**
Implements auto-population of metadata (resource info, http requests, trace and span ids, source location).
Respects auto-population LoggingOptions configuration and WriteOption flag.

### CODE REVIEW

File: google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java [hunk 1]
```diff
@@ -797,6 +799,76 @@ public void write(Iterable<LogEntry> logEntries, WriteOption... options) {
...
+            sharedResourceMetadata == null ? MonitoredResourceUtil.getResource(null, null) : null;
+        final Context context = (new ContextHandler()).getCurrentContext();
+        final ArrayList<LogEntry> populatedLogEntries = Lists.newArrayList();
+        for (LogEntry entry : logEntries) {
+          LogEntry.Builder builder = entry.toBuilder();
+          if (resourceMetadata != null && entry.getResource() == null) {
+            builder.setResource(resourceMetadata);
+          }
+          if (context != null && entry.getHttpRequest() == null) {
+            builder.setHttpRequest(context.getHttpRequest());
```
Comment (losalex) [inline google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java:825]: Should you also check if `context.getHttpRequest() != null`?

**Review by losalex — CHANGES_REQUESTED** [review]
Overall looks good, just asking you to shorten a write function

File: google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java (hunk 1 above)
Comment (minherz) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java:825]: Adding such check does not change the flow. Neither it improves a performance nor security of the execution. Similar logic is applied for if`ed setups of other metadata.

File: google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java [hunk 2]
```diff
@@ -797,6 +799,76 @@ public void write(Iterable<LogEntry> logEntries, WriteOption... options) {
...
+                        .getOrDefault(MonitoredResourceUtil.PORJECTID_LABEL, null);
+              }
+              if (projectId != null) {
+                builder.setTrace(
+                    String.format(RESOURCE_NAME_FORMAT, projectId, context.getTraceId()));
+              } else {
+                builder.setTrace(context.getTraceId());
+              }
+            }
+          }
```
[the remark this answers is not in the document]
Comment (minherz) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java:858]: I will look into it.

[the remark this answers is not in the document]
Comment (losalex) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java:858]: Thanks

**Review by losalex — APPROVED** [review]

File: google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (minherz) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java:858]: moved the trace metadata formatting to a stand-alone method

**Pull request merged**